### ISSUE TITLE
Support receiving incoming nats-messages through millions of low-latency http2 (http3 too?) persistent connections

### ISSUE THREAD

**Issue by dsidirop** [nats-io/nats-server#2772]
## Feature Request

(This is mostly a thought-experiment so please take it with a grain of salt - there might better ways to go about doing this so I'd be happy to stand corrected)

Add built-in support in the nats-server **just** for receiving incoming messages through millions of low-latency http2+ persistent connections (just like with websockets)

I'm referring just to **receiving** nats-messages as http-requests - not responding to them:

![image](https://user-images.githubusercontent.com/7284501/148692543-a663bcb8-4eda-4d90-ad9e-ffdcbbb6bc37.png)

#### Use Case:

Consider that we want to have a platform open to 3rd party clients that use NATS as the subtrate for an event-driven API (the screenshot shows how the client could send a "stop-drone" request)

Currently the nats-server offers WebSocket (later on WebTransport) as a way to let clients publish messages over to NATS.

However the entire cloud ecosystem doesn't have mature enough defensive provisions for WebSocket (yet). That is, all Firewalls, IDS, Auth and Telemetry Systems in major Cloud Platforms are build with HTTP in mind.

For example under the current conditions it is conceivable that an ill-intentioned client might spam-open millions upon millions of new WebSocket-Connections towards the nats-server/nats-cluster and bring the whole thing to its knees or worst. It's a very blunt instrument and yet its very effective as a form of attack.

Same thing holds for the WebSocket rate-limiting aspect (which is not mature enough yet) to fend off ill-intentioneed or mis-behaving clients.

Inconvenient as these shortcomings might be in the defensive perimeter of WebSocket (at least in the way nats deals with it), they are persistent trends that exist and will probably persist for now and the foreseeable future (I would be happy to stand corrected in this aspect of course)

#### Proposed Change:

By enabling the nats-server to have in-grained support for http2+ it can be placed right behind the defensive perimeter of Azure, AWS and enjoy all the benefits of http2+.

The fasthttp project has some jumpstart-ideas as to how we can go about supporting 1.5 million persistent http2-connections and ingest at least 200k RPS with very low-latency:

" Currently fasthttp is successfully used [...] in a production serving up to 200K rps from more than 1.5M concurrent keep-alive connections per physical server "

https://github.com/valyala/fasthttp 
https://github.com/valyala/fasthttp#http-server-performance-comparison-with-nethttp

One could argue that instead of baking http2+ support inside nats-server one could instead write his own custom, front-facing http2+ server and have that http-layer translate incoming http-messages over to nats-messages. That would indeed work but it's problematic because it adds an additional failure point in the engine while also increasing latency and degrading developer-experience.

#### Who Benefits From The Change(s)?

I believe this new arrow in the quiver of nats will speed up its adoption in companies relying on "open" architectures that have SLAs about low-latency, low-jitter and resilience to support fleets of drones created by 3rd-parties. These companies need to ensure that their cloud infrastructure is sufficiently well fortified against the aforementioned kinds of attacks.

#### Alternative Approaches

As already mentioned a company that relies on nats can definately craft its own http2+ translation-layer but this approach is too much of a hassle and negates many of the benefits of "direct connectivity" through NATS as in WebSocket/WebTransport.

**Comment by dsidirop** [comment]
Relevant?

https://github.com/nats-io/nats-server/discussions/2229#discussioncomment-1925325

**Comment by gedw99** [comment]
https://github.com/ripienaar/nats-roundtripper is a decent stop gap ?

**Comment by dsidirop** [comment]
> https://github.com/ripienaar/nats-roundtripper is a decent stop gap ?

Thanks for pointing this out. If I understand correctly it runs as a separate process and it's not built right into nats itself. This is falling short of expectations for reasons detailed in the original post.

**Comment by gedw99** [comment]
Yes it’s a separate process. It’s however compiled into Caddy which acts as a proxy . 

It’s often put in front of nats for protection and ease of deployment.
Caddy is pure golang.
There are a ton of other really useful plugins for caddy too.

I would love webtransport to be ready in NATS too, as I had the same dilemma as you . 

Caddy supports HTTP/3 btw . But not web transport yet .

**Comment by gedw99** [comment]
I want to add that there is a PR that supports http/3 web transport in the golang quic repo. 

Sone are using it , and caddy uses the golang quic lib.

so you could use the or and alter caddy with a PR. Might be worth raising it with caddy team

**Comment by dsidirop** [comment]
> I would love webtransport to be ready in NATS too, as I had the same dilemma as you .

Can you elaborate more? I'm interested to know if WebTransport will help in terms of intergrating NATS with the AWS HTTP-Defenses so as to fend off "spam-connection attacks" and/or rate-limiting.

**Comment by gedw99** [comment]
Yeah it might not. I don't really understand what your trying to do...

**Comment by dsidirop** [comment]
> Yeah it might not. I don't really understand what your trying to do...

I explain it in the diagram. I want to make the communication with NATS to be based on the ingress part in http so as to be able to apply through an intermediate http-infrastructure (Application Load Balancer or whatever) some kind of defense against the following two types of attacks:

1. DoS attacks that are based on spamming traffic towards NATS (if we have rate-limiting on the http level we can defend against this)

2. DoS attacks that are based on spam-opening connections towards NATS (if we have connection-limiting on a per-IP basis we can defend against this as well)

As I already mentioned in the description I'm exploring whether we can use http in the ingress channel of the nats-server exactly for these reasons detailed above.

PS: As a side-thought I discovered that we might be able to rate-limit WebSocket in terms of volume https://github.com/envoyproxy/envoy/issues/13877#issuecomment-945111186   Interesting.

**Comment by gedw99** [comment]
maybe then just put a proxy in front that is customised for NATS ?

**Comment by gedw99** [comment]
hey @dsidirop why not use yomorun as the proxy in front of nats ?

https://github.com/yomorun/yomo

i am using it for the front-end layer in front of nats
